jsk_perception/include/jsk_perception/darknet_object_detection.h

```diff
@@ -0,0 +1,145 @@
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/o2r other materials provided
+ *     with the distribution.
+ *   * Neither the name of the JSK Lab nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Yuki Furuta <furushchev@jsk.imi.i.u-tokyo.ac.jp>
+ *
+ *********************************************************************/
+
+
+#ifndef JSK_PERCEPTION_DARKNET_OBJECT_DETECTION_H__
+#define JSK_PERCEPTION_DARKNET_OBJECT_DETECTION_H__
+
+#include <cstdlib>
+#include <cstring>
+#include <fstream>
+#include <string>
+#include <memory>
+#include <vector>
+#include <boost/filesystem/operations.hpp>
+#include <boost/filesystem/path.hpp>
+#include <boost/shared_ptr.hpp>
+#include <boost/thread.hpp>
+#include <jsk_topic_tools/connection_based_nodelet.h>
+#include <cv_bridge/cv_bridge.h>
+#include <sensor_msgs/image_encodings.h>
+#include <image_transport/image_transport.h>
+#include <image_transport/subscriber_filter.h>
+
+#include <jsk_recognition_msgs/LabeledRectArray.h>
+
+extern "C" {
+#ifdef __cplusplus
+#define ___cplusplus __cplusplus
+#undef __cplusplus
+#endif
+#include <darknet/box.h>
+#include <darknet/network.h>
+#include <darknet/cost_layer.h>
+#include <darknet/region_layer.h>
+#include <darknet/image.h>
+#include <darknet/parser.h>
+#include <darknet/utils.h>
+#ifdef ___cplusplus
+#define __cplusplus ___cplusplus
+#undef ___cplusplus
+#endif
+}
+
+extern "C" image ipl_to_image(IplImage* src);
+
+
+namespace jsk_perception
+{
+
+class AverageArray {
+  float ** arr;
+  bool valid;
+  int n, e, c;
+public:
+  AverageArray(int n, int e) : n(n), e(e), c(0), valid(false) {
```

Comment:
@k-okada This warns order of initializer , to suppress this warning initializer should be

```diff
- AverageArray(int n, int e) : n(n), e(e), c(0), valid(false) {
+ AverageArray(int n, int e) : valid(false), n(n), e(e), c(0) {
```

This will be fixed at the next commit.